### PR TITLE
Added Bin Sort[C]

### PR DESCRIPTION
**Pull request by Lucifer24**
Fixes #372 

#### By submitting this pull request I confirm I've read and complied with the below declarations.

- [x] I have read the [Contribution guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md) and I am confident that my PR reflects them.
- [x] I have followed the [coding guidelines](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#cs) for this project.
- [x] My code follows the [skeleton code structure](https://github.com/iiitv/algos/blob/master/CONTRIBUTING.md#sample).
- [x] This pull request has a descriptive title. For example, `Added {Algorithm/DS name} [{Language}]`, not `Update README.md` or `Added new code`.
- [x] This pull request will be closed if I fail to update it even once in a continuous time span of 7 days.

### CODE REVIEW

**Comment by Monal5031** [comment]
@Lucifer24 Fix the PR description to link issue correctly.

File: bin_sort/bin_sort.c [hunk 1]
```diff
@@ -0,0 +1,58 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+struct bucket {
+	int count;
+	int* value;
+};
+
+int comparator(const void*first, const void* second) {
```
Comment (Monal5031) [inline bin_sort/bin_sort.c]: Space before `first`

File: bin_sort/bin_sort.c [hunk 2]
```diff
@@ -0,0 +1,58 @@
+#include <stdio.h>
```
Comment (Monal5031) [inline bin_sort/bin_sort.c]: Add comments.

File: bin_sort/bin_sort.c [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	}
+	else if (a < b) {
+		return -1;
+	}
+	else {
+		return 1;
+	}
+}
+
+void bucketSort(int array[], int n) {
```
Comment (Monal5031) [inline bin_sort/bin_sort.c]: separate words by underscores, `camelCase` is not followed in C.

File: bin_sort/bin_sort.c [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+		for (b = 0; b < buckets[a].count; b++) {
+			array[c + b] = buckets[a].value[b];
+		}
+		c += buckets[a].count;
+		free(buckets[a].value);
+	}
+}
+
+int main() {
+	int array[12] = {-5, -9, 1000, 1, -10, 0, 2, 3, 5, 4, 1234, 7};
```
Comment (Monal5031) [inline bin_sort/bin_sort.c]: No need to mention 12 explicitly.

File: bin_sort/bin_sort.c [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	int n = sizeof(array) / sizeof(array[0]);
+	int j;
+	bucketSort(array, n);
+	for (j = 0; j < n; j++)
+		printf("%d ", array[j]);
```
Comment (Monal5031) [inline bin_sort/bin_sort.c]: surround this by block here.

File: bin_sort/bin_sort.c [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+#include <stdio.h>
+#include <stdlib.h>
+
+struct bucket {
+	int count;
+	int* value;
+};
+
+int comparator(const void* first, const void* second) {
```
Comment (yashLadha) [inline bin_sort/bin_sort.c]: Why to explicitly cat into `int *` When u can pass them in the function declaration??

File: bin_sort/bin_sort.c [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+int main() {
+	int array[] = {-5, -9, 1000, 1, -10, 0, 2, 3, 5, 4, 1234, 7};
+	int n = sizeof(array) / sizeof(array[0]);
+	int j;
+	bucket_sort(array, n);
+	for (j = 0; j < n; j++) {
+		printf("%d ", array[j]);
+	}
+	return 0;
+}
```
Comment (yashLadha) [inline bin_sort/bin_sort.c]: Add new at EOF

File: bin_sort/bin_sort.c [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+	}
+}
+
+int main() {
+	int array[] = {-5, -9, 1000, 1, -10, 0, 2, 3, 5, 4, 1234, 7};
+	int n = sizeof(array) / sizeof(array[0]);
+	int j;
+	bucket_sort(array, n);
+	for (j = 0; j < n; j++) {
+		printf("%d ", array[j]);
```
Comment (yashLadha) [inline bin_sort/bin_sort.c]: Add endl after alter whole output

File: bin_sort/bin_sort.c (hunk 7 above)
Comment (aashutoshrathi) [inline bin_sort/bin_sort.c]: @yashLadha There is newline at EOF, it can't be seen here

Comment (yashLadha) [inline bin_sort/bin_sort.c]: Are u using and IDE. Because it is seen in other codes, if it is present their.

File: bin_sort/bin_sort.c (hunk 6 above)
Comment (Lucifer24) [inline bin_sort/bin_sort.c]: Can you explain it ?

Comment (yashLadha) [inline bin_sort/bin_sort.c]: You are type casting a `void *` into an `int *`. Instead of that pass `int *` in the definition itself.

Comment (yashLadha) [inline bin_sort/bin_sort.c]: This will prevent the overhead of typecasting.

**Pull request closed without merging**